qiskit/aqua/algorithms/phase_estimators/phase_estimator.py

```diff
@@ -0,0 +1,208 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""The Quantum Phase Estimation Algorithm."""
+
+
+from typing import Optional, Union
+import numpy
+from qiskit.circuit import QuantumCircuit
+import qiskit
+from qiskit.circuit.library import PhaseEstimation
+from qiskit.circuit.classicalregister import ClassicalRegister
+from qiskit.providers import BaseBackend
+from qiskit.aqua import QuantumInstance
+from qiskit.aqua.algorithms import QuantumAlgorithm
+from qiskit.aqua.utils import get_subsystem_density_matrix
+from qiskit.result import Result
+from .phase_estimator_result import PhaseEstimatorResult, _sort_phases
+
+
+class PhaseEstimator(QuantumAlgorithm):
+    """Run the Quantum Phase Estimation algorithm.
+
+    This runs a version of QPE with a multi-qubit register for reading the phase [1]. The main
+    inputs are the number of qubits in the phase-reading register, a state preparation circuit to
+    prepare an input state, and either
+    1) A unitary that will act on the the input state, or
+    2) A quantum-phase-estimation circuit in which the unitary is already embedded.
+    In case 1), an instance of `qiskit.circuit.PhaseEstimation`, a QPE circuit, containing the input
+    unitary will be constructed. After construction, the QPE circuit is run on a backend via the
+    `run` method, and the frequencies or counts of the phases represented by bitstrings are
+    recorded. The results are returned as an instance of
+    qiskit.algorithms.phase_estimator_result.PhaseEstimatorResult.
+
+    If the input state is an eigenstate of the unitary, then in the ideal case, all probability is
+    concentrated on the bitstring corresponding to the eigenvalue of the input state. If the input
+    state is a superposition of eigenstates, then each bitstring is measured with a probability
+    corresponding to its weight in the superposition. In addition, if the phase is not representable
+    exactly by the phase-reading register, the probability will be spread across bitstrings, with an
+    amplitude that decreases with distance from the bitstring most closely approximating the phase.
+
+    **Reference:**
+
+    [1]: Michael A. Nielsen and Isaac L. Chuang. 2011.
+         Quantum Computation and Quantum Information: 10th Anniversary Edition (10th ed.).
+         Cambridge University Press, New York, NY, USA.
+
+    """
+
+    def __init__(self,
+                 num_evaluation_qubits: int,
+                 unitary: Optional[QuantumCircuit] = None,
+                 pe_circuit: Optional[QuantumCircuit] = None,
+                 num_unitary_qubits: Optional[int] = None,
+                 state_preparation: Optional[QuantumCircuit] = None,
```

Comment:
Could you move this argument above `pe_circuit`? This way, `unitary` and `state_preparation` are grouped such that they could be used in an unnamed way.